### PR TITLE
RISC-V support

### PR DESCRIPTION
**Pull request by psychocoderHPC**
This commit is providing nessesare changes to cross-compile PIConGPU for RISC-V.

- refactor CMake and provide an workaround required to link static MPI libraries during cross compilation
- add short cross compile section to the documentation
- provide GNU and Clang CMake toolchain file

### CODE REVIEW

File: share/ci/compiler_hipcc.yml [hunk 1]
```diff
@@ -10,6 +10,8 @@
     #  "Could NOT find MPI (missing: MPI_C_FOUND MPI_CXX_FOUND)"
     PIC_CMAKE_ARGS: "-DMPI_CXX_WORKS=ON -DMPI_CXX_VERSION=0"
     # ISAAC is not working with HIP
```
Comment (PrometheusPi) [inline share/ci/compiler_hipcc.yml:12]: not part of the PR: Is this still true?

File: include/pmacc/PMaccConfig.cmake [hunk 2]
```diff
@@ -288,13 +287,33 @@ endif()
...
 find_package(MPI REQUIRED)
-target_include_directories(pmacc PUBLIC ${MPI_C_INCLUDE_PATH})
-target_link_libraries(pmacc PUBLIC ${MPI_C_LIBRARIES})
+target_link_libraries(pmacc PUBLIC MPI::MPI_CXX)
+target_link_libraries(pmacc_filesystem INTERFACE MPI::MPI_CXX)
+
+if( CMAKE_TRY_COMPILE_TARGET_TYPE STREQUAL "STATIC_LIBRARY" AND CMAKE_EXE_LINKER_FLAGS)
+    # Workaround for linker issues when linking static MPI libraries.
+    # Because of CMAKE_TRY_COMPILE_TARGET_TYPE CMake is providing the statics libraries before the object file from
+    # `add_executable` therefore MPI symbols can not be resolved. Linking the linker flaks to the target again will
```
Comment (PrometheusPi) [inline include/pmacc/PMaccConfig.cmake:296]:
```suggestion
    # `add_executable` therefore MPI symbols can not be resolved. Linking the linker flags to the target again will
```

Is this a typo?

File: include/pmacc/CMakeLists.txt [hunk 3]
```diff
@@ -71,7 +71,7 @@ get_filename_component(PMACC_ROOT_DIR "${PMACC_ROOT_DIR}" ABSOLUTE)
 # PMacc
 ################################################################################
 
-find_package(PMacc REQUIRED CONFIG PATHS ${CMAKE_CURRENT_SOURCE_DIR})
+find_package(PMacc REQUIRED CONFIG PATHS test)
```
Comment (PrometheusPi) [inline include/pmacc/CMakeLists.txt]: are `${CMAKE_CURRENT_LIST_DIR} == ${CMAKE_CURRENT_SOURCE_DIR}) == test`?

File: include/picongpu/CMakeLists.txt [hunk 4]
```diff
@@ -315,6 +315,10 @@ if(PIC_SEARCH_openPMD)
     # find openPMD installation
     find_package(openPMD 0.14.3 CONFIG COMPONENTS MPI)
     if(openPMD_FOUND)
+        # HDF5 and ADIOS is requires linking against C libraries but this is not correctly
```
Comment (PrometheusPi) [inline include/picongpu/CMakeLists.txt]:
```suggestion
        # HDF5 and ADIOS require linking against C libraries but this is not correctly
```

File: include/mpiInfo/CMakeLists.txt [hunk 5]
```diff
@@ -80,20 +80,29 @@ set(CMAKE_CXX_STANDARD 17)
...
+# Find Threads
+################################################################################
+
+# Allow users to override the "-pthread" preference.
+if(NOT THREADS_PREFER_PTHREAD_FLAG)
+     set(THREADS_PREFER_PTHREAD_FLAG TRUE)
+endif()
+
+# WARNING: in case were we cross compile we need to add '-DDMPI_CXX_WORKS=ON` and
+# provide include flags and linker flags manual.
```
Comment (PrometheusPi) [inline include/mpiInfo/CMakeLists.txt]:
```suggestion
# provide include flags and linker flags manually.
```

File: include/mpiInfo/CMakeLists.txt [hunk 6]
```diff
@@ -116,7 +125,20 @@ add_executable(mpiInfo
...
+target_link_libraries(mpiInfo PRIVATE ${MPIINFO_LIBS} Threads::Threads)
+if(RT_LIBRARY AND NOT APPLE)
+    # MPI runtime libraries
+    target_link_libraries(mpiInfo PRIVATE ${RT_LIBRARY})
+endif()
+target_link_libraries(mpiInfo PRIVATE MPI::MPI_CXX)
+
+if( CMAKE_TRY_COMPILE_TARGET_TYPE STREQUAL "STATIC_LIBRARY" AND CMAKE_EXE_LINKER_FLAGS)
+    # Workaround for linker issues when linking static MPI libraries.
+    # Because of CMAKE_TRY_COMPILE_TARGET_TYPE CMake is providing the statics libraries before the object file from
```
Comment (PrometheusPi) [inline include/mpiInfo/CMakeLists.txt]:
```suggestion
    # Because of CMAKE_TRY_COMPILE_TARGET_TYPE CMake providing the statics libraries before the object file from
```

File: include/mpiInfo/CMakeLists.txt [hunk 7]
```diff
@@ -116,7 +125,20 @@ add_executable(mpiInfo
...
+    # `add_executable` therefore MPI symbols can not be resolved. Linking the linker flaks to the target again will
```
Comment (PrometheusPi) [inline include/mpiInfo/CMakeLists.txt]:
```suggestion
    # `add_executable`, MPI symbols can not be resolved. Linking the linker flags to the target again will
```
Is `flaks` a typo?

File: docs/source/usage/crosscompile.rst [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+.. code-block:: bash
+
+   # set MPI_ROOT to the root MPI directory
+   export LDFLAGS="-L$MPI_ROOT/lib -lmpi -lompitrace -lopen-rte -lopen-pal -ldl -lpthread -lutil -lrt"
+   # compile for 64bit RISC-V with double floating point support
+   export CXXFLAGS="-I$MPI_ROOT/include -pthread -march=rv64gc -mabi=lp64d"
+   export CFLAGS=$CXXFLAGS
+
+To be able to cross compile you should point to the CMake toolchain file shipped together with PIConGPU.
+Depending on your environment please select the Clang or GNU toolchain.
```
Comment (PrometheusPi) [inline docs/source/usage/crosscompile.rst]:
```suggestion
Depending on your environment, please select the Clang or GNU toolchain.
```

File: docs/source/usage/crosscompile.rst [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+The execution backend is provided in this example explicit.
```
Comment (PrometheusPi) [inline docs/source/usage/crosscompile.rst]:
```suggestion
The execution backend is provided in this example explicitly.
```

File: docs/source/usage/crosscompile.rst [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+   export RISCV_GNU_INSTALL_ROOT="$(dirname $(which riscv64-unknown-linux-gnu-gcc))/.."
+   export RISCV_CLANG_INSTALL_ROOT="$(dirname $(which clang))/.."
+
+.. code-block:: bash
+
+   # ``omp2b`` for OpenMP and ``serial`` for serial execution on one core
+   pic-build -b omp2b -c"-DMPI_CXX_WORKS=ON  -DCMAKE_TOOLCHAIN_FILE=<PATH_TO_PICONGPU_SRC>/cmake/riscv64-gnu.toolchain.cmake"
+
+After PICoNGPU is compiled you can interactively join to the RISC-V compute node and test if you compield for the right target architecture ``./bin/picongpu --help``.
```
Comment (PrometheusPi) [inline docs/source/usage/crosscompile.rst]:
```suggestion
After PIConGPU is compiled you can interactively join to the RISC-V compute node and test if you compiled for the right target architecture ``./bin/picongpu --help``.
```

File: cmake/riscv64-gnu.toolchain.cmake [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+set(GNU_TARGET_TRIPLE riscv64-unknown-linux-gnu)
+
+set(CMAKE_C_COMPILER ${RISCV_GNU_INSTALL_ROOT}/bin/riscv64-unknown-linux-gnu-gcc)
+set(CMAKE_C_COMPILER_TARGET ${GNU_TARGET_TRIPLE})
+set(CMAKE_CXX_COMPILER ${RISCV_GNU_INSTALL_ROOT}/bin/riscv64-unknown-linux-gnu-g++)
+set(CMAKE_CXX_COMPILER_TARGET ${GNU_TARGET_TRIPLE})
+set(CMAKE_ASM_COMPILER ${RISCV_GNU_INSTALL_ROOT}/bin/riscv64-unknown-linux-gnu-as)
+set(CMAKE_ASM_COMPILER_TARGET ${GNU_TARGET_TRIPLE})
+
+# Don't run the linker on compiler check
```
Comment (PrometheusPi) [inline cmake/riscv64-gnu.toolchain.cmake]: optionally:
```suggestion
# Don't run the linker on compiler check.
```
And I do not understand what this means 😅

File: cmake/riscv64-clang.toolchain.cmake [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+set(CMAKE_C_COMPILER ${RISCV_CLANG_INSTALL_ROOT}/bin/clang)
+set(CMAKE_C_COMPILER_TARGET ${CLANG_TARGET_TRIPLE})
+set(CMAKE_CXX_COMPILER ${RISCV_CLANG_INSTALL_ROOT}/bin/clang++)
+set(CMAKE_CXX_COMPILER_TARGET ${CLANG_TARGET_TRIPLE})
+set(CMAKE_ASM_COMPILER ${RISCV_CLANG_INSTALL_ROOT}/bin/clang)
+set(CMAKE_ASM_COMPILER_TARGET ${CLANG_TARGET_TRIPLE})
+
+
+# Don't run the linker on compiler check
```
Comment (PrometheusPi) [inline cmake/riscv64-clang.toolchain.cmake]: Optionally: 
```suggestion
# Don't run the linker on compiler check.
```

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
I checked for typos and the correct use of the new append syntax. 
I am the wrong person to validate RISC-V cmake setups.

File: share/ci/compiler_hipcc.yml (hunk 1 above)
Comment (psychocoderHPC) [reply share/ci/compiler_hipcc.yml:12]: Should work and should be checked separately.

File: include/pmacc/CMakeLists.txt (hunk 3 above)
Comment (psychocoderHPC) [reply include/pmacc/CMakeLists.txt]: no left over from a test, good catch

File: cmake/riscv64-gnu.toolchain.cmake (hunk 11 above)
Comment (psychocoderHPC) [reply cmake/riscv64-gnu.toolchain.cmake]: changed the sentence

File: share/ci/compiler_hipcc.yml (hunk 1 above)
Comment (PrometheusPi) [reply share/ci/compiler_hipcc.yml:12]: Okay - I opened #4714

File: include/pmacc/CMakeLists.txt (hunk 3 above)
Comment (PrometheusPi) [reply include/pmacc/CMakeLists.txt]: Is it still "test"? (At least I seee "test" in the code).

**Review by PrometheusPi — DISMISSED** [review]
Just one question.

**Review by PrometheusPi — APPROVED** [review]
